File: services/geminiService.ts

import { Difficulty } from "../types";

const DEEPSEEK_API_URL = "https://api.deepseek.com/v1/chat/completions";

export interface ChainData {
  chain: string[];
  explanations: string[];
}

const complexityMap = {
  'EASY': 'very simple common two-word phrases (e.g., COLD WATER, ICE CREAM)',
  'MEDIUM': 'common idioms and collocations (e.g., SOCIAL MEDIA, FIELD TRIP, HIGH SCHOOL)',
  'HARD': 'complex idioms and abstract connections (e.g., COLD FEET, SILVER LINING, NIGHT OWL)'
};

export const generateWordChainData = async (difficulty: Difficulty): Promise<ChainData> => {
  try {
    const response = await fetch(DEEPSEEK_API_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${process.env.DEEPSEEK_API_KEY}`,
      },
      body: JSON.stringify({
        model: "deepseek-chat",
        messages: [{
          role: "user",
          content: `Generate an interesting word chain of 9 English words.
Difficulty level: ${difficulty}. Use ${complexityMap[difficulty]}.
Each consecutive pair MUST form a common two-word phrase, idiom, or collocation.
CRITICAL RULE: DO NOT split single compound words into two parts (e.g., do not split "SUNFLOWER" into "SUN" and "FLOWER").
Instead, use separate words that form a phrase, like "SUN" and "RAYS" (Sun rays).
For each pair (8 pairs total), provide a very brief one-sentence explanation of what that phrase means.
Respond with valid JSON only, no markdown, in this exact format:
{"chain":["WORD1","WORD2","WORD3","WORD4","WORD5","WORD6","WORD7","WORD8","WORD9"],"explanations":["WORD1 WORD2: explanation.","WORD2 WORD3: explanation.","WORD3 WORD4: explanation.","WORD4 WORD5: explanation.","WORD5 WORD6: explanation.","WORD6 WORD7: explanation.","WORD7 WORD8: explanation.","WORD8 WORD9: explanation."]}`,
        }],
        response_format: { type: "json_object" },
      }),
    });

    if (!response.ok) throw new Error(`DeepSeek error: ${response.status}`);

    const json = await response.json();
    const data = JSON.parse(json.choices[0].message.content);

    if (data.chain?.length >= 9 && data.explanations?.length >= 8) {
      return {
        chain: data.chain.map((w: string) => w.toUpperCase().trim()),
        explanations: data.explanations,
      };
    }
    throw new Error("Invalid data format");
  } catch (error) {
    console.error("Error generating chain:", error);
    // Fallback data for stability
    return {
      chain: ["ICE", "CREAM", "SODA", "WATER", "BOTTLE", "CAP", "TAIN", "COOK", "BOOK"],
      explanations: [
        "ICE CREAM: A frozen dessert.",
        "CREAM SODA: A sweet carbonated drink.",
        "SODA WATER: Carbonated water.",
        "WATER BOTTLE: A container for holding water.",
        "BOTTLE CAP: A closure for a bottle.",
        "CAPTAIN COOK: A famous explorer.",
        "COOK BOOK: A book containing recipes.",
        "BOOK CLUB: A group of people who meet to discuss books.",
      ],
    };
  }
};

File: WordLink/Services/GeminiService.swift
import Foundation

// MARK: - OpenAI API Response Types
private struct OpenAIResponse: Decodable {
    let choices: [Choice]

    struct Choice: Decodable {
        let message: Message
    }

    struct Message: Decodable {
        let content: String
    }
}

private struct PairResponse: Decodable {
    let w1: String
    let w2: String
    let explanation: String
}

private struct ChainResponse: Decodable {
    let chain: [String]
    let pairs: [PairResponse]
}

// MARK: - AI Service (OpenAI backend)
final class GeminiService {
    static var apiKey: String = "REMOVED"

    private static let endpoint = "https://api.openai.com/v1/chat/completions"

    private static let fallbackData = ChainData(
        chain: ["HIGH", "SCHOOL", "BUS", "STOP", "SIGN", "LANGUAGE", "BARRIER", "REEF", "KNOT"],
        explanations: [
            "HIGH SCHOOL: A secondary school for students aged roughly 14–18.",
            "SCHOOL BUS: A vehicle that transports students to and from school.",
            "BUS STOP: A designated place where passengers board or exit a bus.",
            "STOP SIGN: A red octagonal road sign requiring vehicles to halt.",
            "SIGN LANGUAGE: A visual language using hand shapes and movements.",
            "LANGUAGE BARRIER: Difficulty communicating due to different languages.",
            "BARRIER REEF: A coral reef running parallel to a coastline.",
            "REEF KNOT: A simple, symmetrical knot used to bind two ropes together."
        ]
    )

    static func generateWordChain(difficulty: Difficulty) async -> ChainData {
        let complexityMap: [Difficulty: String] = [
            .easy: "very simple common two-word phrases (e.g., COLD WATER, ICE CREAM)",
            .medium: "common idioms and collocations (e.g., SOCIAL MEDIA, FIELD TRIP, HIGH SCHOOL)",
            .hard: "complex idioms and abstract connections (e.g., COLD FEET, SILVER LINING, NIGHT OWL)"
        ]

        let prompt = """
        Generate a word chain of exactly 9 English words.
        Difficulty level: \(difficulty.rawValue). Use \(complexityMap[difficulty] ?? "").

        STRICT RULES — read carefully:
        1. Every word in the chain must be a real, standalone English dictionary word.
        2. Each consecutive pair must form a common TWO-WORD PHRASE — two separate words used together as a phrase (e.g. "high school", "bus stop", "sign language", "cold water", "social media").
        3. The pair must NOT merge into a single compound word:
           - FORBIDDEN: BLACK + BIRD → "blackbird" (one merged word)
           - FORBIDDEN: WATER + FALL → "waterfall" (one merged word)
           - FORBIDDEN: CAP + TAIN → "captain" (not real words joined)
           - ALLOWED: COLD + WATER → "cold water" (genuine two-word phrase)
           - ALLOWED: HIGH + SCHOOL → "high school" (genuine two-word phrase)
        4. Ask yourself: "Is this phrase normally written as two separate words?" If yes, allowed. If one word, forbidden.

        For each of the 8 consecutive pairs, provide a "pairs" entry with the exact words and explanation.
        Respond with JSON only:
        {"chain": ["WORD1","WORD2","WORD3","WORD4","WORD5","WORD6","WORD7","WORD8","WORD9"], "pairs": [{"w1":"WORD1","w2":"WORD2","explanation":"WORD1 WORD2: one sentence definition."},{"w1":"WORD2","w2":"WORD3","explanation":"WORD2 WORD3: one sentence definition."},{"w1":"WORD3","w2":"WORD4","explanation":"WORD3 WORD4: one sentence definition."},{"w1":"WORD4","w2":"WORD5","explanation":"WORD4 WORD5: one sentence definition."},{"w1":"WORD5","w2":"WORD6","explanation":"WORD5 WORD6: one sentence definition."},{"w1":"WORD6","w2":"WORD7","explanation":"WORD6 WORD7: one sentence definition."},{"w1":"WORD7","w2":"WORD8","explanation":"WORD7 WORD8: one sentence definition."},{"w1":"WORD8","w2":"WORD9","explanation":"WORD8 WORD9: one sentence definition."}]}
        """

        let requestBody: [String: Any] = [
            "model": "gpt-4o-mini",
            "messages": [
                ["role": "user", "content": prompt]
            ],
            "response_format": ["type": "json_object"]
        ]

        guard let url = URL(string: endpoint),
              let bodyData = try? JSONSerialization.data(withJSONObject: requestBody) else {
            return fallbackData
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = bodyData

        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return fallbackData
            }

            let openAIResponse = try JSONDecoder().decode(OpenAIResponse.self, from: data)
            guard let content = openAIResponse.choices.first?.message.content,
                  let jsonData = content.data(using: .utf8) else {
                return fallbackData
            }

            let chainResponse = try JSONDecoder().decode(ChainResponse.self, from: jsonData)
            let upperChain = chainResponse.chain.map { $0.uppercased().trimmingCharacters(in: .whitespaces) }

            if upperChain.count >= 9 && chainResponse.pairs.count >= 8 {
                // Match each pair by w1/w2 to guarantee correct explanation for each slot
                var orderedExplanations: [String] = []
                for i in 0..<8 {
                    let w1 = upperChain[i]
                    let w2 = upperChain[i + 1]
                    if let match = chainResponse.pairs.first(where: {
                        $0.w1.uppercased().trimmingCharacters(in: .whitespaces) == w1 &&
                        $0.w2.uppercased().trimmingCharacters(in: .whitespaces) == w2
                    }) {
                        orderedExplanations.append(match.explanation)
                    } else {
                        // Fallback: use positional index if matching fails
                        orderedExplanations.append(chainResponse.pairs.indices.contains(i) ? chainResponse.pairs[i].explanation : "")
                    }
                }
                return ChainData(chain: upperChain, explanations: orderedExplanations)
            }
            return fallbackData
        } catch {
            print("OpenAI API error: \(error)")
            return fallbackData
        }
    }
}
